### PR TITLE
Transition scheduled flow runs immediately to cancelled instead of cancelling

### PR DESCRIPTION
**Pull request by peytonrunyan**
Closes https://github.com/PrefectHQ/prefect/issues/8194

This PR adds a flow orchestration rule that rejects attempts to transition scheduled flow runs to a `CANCELLING` state, and instead forces them into a `CANCELLED` state. This keeps the agent from attempting to shutdown infrastructure that does not exist for the scheduled flow run (which is what happens for flow runs in a `CANCELLING` state). 

I went with this approach over the other approach that @madkinsz offered because it keeps orchestration logic on the API.

### Example

Before bug fix:
https://www.loom.com/share/b708257a5fe74433853de084e96faf8c

After Bug Fix:
https://www.loom.com/share/60e60d54de47437b8ad72076cfa4e443
### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/orion/orchestration/core_policy.py [hunk 1]
```diff
@@ -811,3 +812,27 @@ async def before_transition(
             ),
         )
         return
+
+
+class HandleCancellingScheduledFlowRuns(BaseOrchestrationRule):
+    """
+    Rejects transitions from Scheduled to Cancelling and set the state to Cancelled.
```
Comment (peytonrunyan) [inline src/prefect/orion/orchestration/core_policy.py:819]:
```suggestion
    Rejects transitions from Scheduled to Cancelling and instead sets the state to Cancelled.
```

**Comment by peytonrunyan** [comment]
Hmmmm, this may be slightly complicated by `AwaitingRetry` being a `SCHEDULED` state

File: src/prefect/orion/orchestration/core_policy.py [hunk 2]
```diff
@@ -811,3 +812,31 @@ async def before_transition(
             ),
         )
         return
+
+
+class HandleCancellingScheduledFlowRuns(BaseOrchestrationRule):
+    """Rejects transitions from Scheduled and Late to Cancelling and set the state to Cancelled.
```
Comment (peytonrunyan) [inline src/prefect/orion/orchestration/core_policy.py:818]:
```suggestion
    """Rejects transitions from Scheduled and Late to Cancelling and sets the state to Cancelled.
```

Comment (peytonrunyan) [reply src/prefect/orion/orchestration/core_policy.py:818]:
```suggestion
    """Rejects transitions from Scheduled to Cancelling and sets the state to Cancelled if the flow run has no associated infrastructure process ID.
```

File: src/prefect/orion/orchestration/core_policy.py [hunk 3]
```diff
@@ -42,6 +42,7 @@ def priority():
         return [
             HandleFlowTerminalStateTransitions,
             HandleCancellingStateTransitions,
+            HandleCancellingScheduledFlowRuns,
```
[the remark this answers is not in the document]
Comment (anticorrelator) [reply src/prefect/orion/orchestration/core_policy.py:45]: We probably can put it into `HandleCancellingStateTransitions` for brevity

File: src/prefect/orion/orchestration/core_policy.py [hunk 4]
```diff
@@ -811,3 +812,32 @@ async def before_transition(
...
         )
         return
+
+
+class HandleCancellingScheduledFlowRuns(BaseOrchestrationRule):
+    """Rejects transitions from Scheduled to Cancelling, and sets the state to Cancelled, if
+    the flow run has no associated infrastructure process ID.
+
+    The `Cancelling` state is used to clean up infrastructure. If there is not infrastructure
+    to clean up, we can transition directly to `Cancelled`. Runs that are `AwaitingRetry` are
```
Comment (anticorrelator) [inline src/prefect/orion/orchestration/core_policy.py:822]: Does this comment indicate we want to exclude `AwaitingRetry` named states from this short-circuit?

File: tests/cli/test_flow_run.py [hunk 5]
```diff
@@ -221,15 +221,15 @@ class TestCancelFlowRun:
     @pytest.mark.parametrize(
         "state",
         [
-            Scheduled,
             Running,
-            Late,
             Pending,
-            AwaitingRetry,
```
Comment (anticorrelator) [inline tests/cli/test_flow_run.py:228]: As above, should `AwaitingRetry` named states still be put in a `Cancelling` state?

File: src/prefect/orion/orchestration/core_policy.py (hunk 4 above)
Comment (peytonrunyan) [reply src/prefect/orion/orchestration/core_policy.py:822]: It's currently just a context comment. In theory if there is somehow an AwaitingRetry flow run without an infra PID, we should jump it straight to cancelled too

File: tests/cli/test_flow_run.py (hunk 5 above)
Comment (peytonrunyan) [reply tests/cli/test_flow_run.py:228]: As long as they don't have an infra PID

File: tests/orion/orchestration/test_core_policy.py [hunk 6]
```diff
@@ -2599,3 +2600,92 @@ async def test_rejects_cancelled_cancelling_to_anything_but_cancelled(
...
+            session,
+            "flow",
+            *intended_transition,
+        )
+
+        async with HandleCancellingScheduledFlowRuns(ctx, *intended_transition) as ctx:
+            await ctx.validate_proposed_state()
+
+        assert ctx.response_status == SetStateStatus.REJECT
+        assert ctx.validated_state_type == states.StateType.CANCELLED
```
Comment (serinamarie) [inline tests/orion/orchestration/test_core_policy.py:2676]: Are lines 2663-2676 needed? How is this different than lines 2617-2627?

File: tests/orion/orchestration/test_core_policy.py [hunk 7]
```diff
@@ -2599,3 +2600,92 @@ async def test_rejects_cancelled_cancelling_to_anything_but_cancelled(
...
+            *intended_transition,
+        )
+
+        async with HandleCancellingScheduledFlowRuns(ctx, *intended_transition) as ctx:
+            await ctx.validate_proposed_state()
+
+        assert ctx.response_status == SetStateStatus.ACCEPT
+        assert ctx.validated_state_type == states.StateType.CANCELLING
+
+    async def test_accepts_cancelling_flow_run_with_pid(
```
Comment (serinamarie) [inline tests/orion/orchestration/test_core_policy.py:2655]:
```suggestion
    async def test_accepts_cancelling_scheduled_flow_run_and_sets_to_cancelling(
```

Would this match better with the other test name?

File: tests/orion/orchestration/test_core_policy.py [hunk 8]
```diff
@@ -2599,3 +2600,92 @@ async def test_rejects_cancelled_cancelling_to_anything_but_cancelled(
...
+        self,
+        session,
+        initialize_orchestration,
+    ):
+        """Flow runs awaiting retry should still go into a cancelling state even though they.
+        are technically "Scheduled".
```
Comment (serinamarie) [inline tests/orion/orchestration/test_core_policy.py:2661]:
```suggestion
        """Flow runs awaiting retry should still go into a cancelling state as they have an associated pid, even though they are technically "Scheduled".
```

**Comment by peytonrunyan** [comment]
Spoke with Dustin about it and we decided to keep the rules separate and rename them for clarity. I'm going to keep this as a draft until things are running in Cloud.

File: tests/orion/orchestration/test_core_policy.py (hunk 6 above)
Comment (peytonrunyan) [reply tests/orion/orchestration/test_core_policy.py]: This was mostly a behavior sanity check. "This is behavior at point X" and then "This is behavior at point X + <change>". It is 100% duplicating the actual test, and I'm never quite sure how correct it is to do something like. I'll ask in the engineering channel.

File: src/prefect/orion/orchestration/core_policy.py [hunk 9]
```diff
@@ -826,13 +829,14 @@ async def before_transition(
...
-            ),
-        )
-        return
+
+        if not context.run.infrastructure_pid:
+            await self.reject_transition(
+                state=states.Cancelled(),
+                reason=(
+                    "Scheduled flows are transitioned immediately to Cancelled because there is no "
+                    "infrastructure to shut down."
```
Comment (madkinsz) [inline src/prefect/orion/orchestration/core_policy.py:841]: This should probably just read: "Scheduled flow run has no infrastructure to terminate"

**Comment by madkinsz** [comment]
@serinamarie Can you take this over?

**Comment by cumulus-nrussell** [comment]
It would also be great if automations that are designed to cancel a flow would immediately switch to "Cancelled". I have one such automation where the flows just get stuck in "Cancelling" and take up my concurrency limits.

**Comment by madkinsz** [comment]
@cumulus-nrussell we cannot immediately transition them to "Cancelled" if they're not yet cancelled. It is by-design that a flow run does not release its concurrency slot until it is cancelled, otherwise, it your limits may be exceeded. This pull request is specifically for cancelling runs in a SCHEDULED state though and is likely not relevant to your issue.

**Review by serinamarie — APPROVED** [review]

**Pull request merged**